cachito/workers/pkg_managers/npm.py

```diff
@@ -293,65 +293,187 @@ def get_package_name_from_path(package_path: str) -> str:
+    If the dependency cannot be resolved, return None.
     """
-    path = Path(package_path)
-    dep_paths = [path / "node_modules" / dep_name]
-    for parent_path in path.parents:
-        if parent_path == Path(""):
-            dep_paths.append(Path("node_modules", dep_name))
-        elif parent_path.stem == "node_modules" and parent_path.parent != Path(""):
-            dep_paths.append(parent_path / dep_name)
+    if dep_name in node.children:
+        return node.children[dep_name].package
+
+    if not node.parent:
+        log.warning(
+            f"Cachito was unable to resolve dependency {dep_name} in the package tree. "
+            "It may be an optional peerDependency that isn't included in package-lock.json"
+        )
+        return None
 
-    return [path.as_posix() for path in dep_paths]
+    return _resolve_node_dependency(node.parent, dep_name)
+
+
+def _get_v2_package_tree(paths_to_packages: dict[str, Package]) -> PackageTreeNode:
+    """Return the root PackageTreeNode for the `packages` in a v2+ package-lock.json file.
+
+    A v2+ package-lock.json file contains a flat `packages` dict where the keys are the paths
+    to a given package. Use the paths to generate a tree structure so that we can later
+    determine which packages depend on other packages and do non-registry dependency replacements.
+    """
+    root_node = PackageTreeNode()
+    paths_to_nodes = {
+        Path(path): PackageTreeNode(package) for path, package in paths_to_packages.items()
+    }
+
+    for path, package_node in paths_to_nodes.items():
+        package = package_node.package
+        if package is None:
+            raise CachitoError(
+                (
+                    "Cachito encountered an error while constructing the npm package tree. "
+                    f"Package node {package_node} has no associated Package. "
+                    "This should never happen."
+                )
+            )
+        parent_node = (
+            _get_parent_node(path, root_node, paths_to_nodes)
+            if "node_modules" in path.parts
+            else _get_fsparent_node(path, root_node, paths_to_nodes)
+        )
+        if parent_node == root_node:
+            package.is_top_level = True
+        package_node.parent = parent_node
+        parent_node.children[package.alias or package.name] = package_node
+
+    return root_node
+
+
+def _get_fsparent_node(
+    path: Path, root_node: PackageTreeNode, paths_to_nodes: dict[Path, PackageTreeNode]
+) -> PackageTreeNode:
+    """Return the PackageTreeNode for the Package that is the fsparent for the given path.
+
+    https://github.com/npm/cli/blob/latest/workspaces/arborist/docs/parent.md#fsparent
+    """
+    if path == Path(""):
+        return root_node
+
+    for parent in path.parents:
+        if parent == Path(""):
+            return root_node
+        if parent in paths_to_nodes:
+            return paths_to_nodes[parent]
```

Comment:
For my understanding, is it important that the parent node should be `root_node` rather than the main package node?

```python
{
  Path(""): PackageTreeNode(main_package)
)
```
^ this one would otherwise match everything, but here we specifically make sure to never match it and return root_node instead